### ISSUE TITLE
Submodules of git dependencies are not fetched

### ISSUE THREAD

**Issue by ntucker** [yarnpkg/yarn#1488]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
Having a git dependency with a submodule results in the submodule being an empty directory

**If the current behavior is a bug, please provide the steps to reproduce.**

1. Setup NPM package at a github url
1. Add submodule to said package
1. Add dependency in package.json for local project that points to the github url of said package
1. Yarn install
1. Check node_modules directory for that package and see that submodule folder is empty.

**What is the expected behavior?**
Submodule directories should have the contents (the cloning process should run submodule init and submodule update)

**Please mention your node.js, yarn and operating system version.**
Node: 7.0.0
Yarn: 0.16.1
OS: Ubuntu 14.04

**Comment by jokeyrhyme** [comment]
Here's the discussion for when they added this to `npm`: https://github.com/npm/npm/issues/1876

**Comment by hector** [comment]
+1

Some npm packages won't work without this feature.

**Comment by DeDuckProject** [comment]
Moreover, the submodules get **deleted** if they are in there rightful place before running `yarn`

**Comment by hector** [comment]
I look forward to start using yarn but I cannot switch until this gets fixed :(

**Comment by leggomuhgreggo** [comment]
+1

**Comment by rogerkerse** [comment]
Any news on it?

**Comment by gedgaudasnikita** [comment]
While i hope this issue is getting some research invested into it, this placeholder workaround, that I despise with all my soul, might serve as a temporary solution for all of us waiting 

In package.json of a module being fetched, add a following script
`"postinstall": "grep url .gitmodules | sed 's/.*= //' | while read url; do git clone $url; done",`

**Comment by Qix-** [comment]
So here's the synopsis of the problem:

The Git utilities included with Yarn seem to do a `git clone` and then `git archive` to archive the repository. I've modified the `clone` command to include `--depth=1 --recursive`, but the problem is with `git archive` - it's been identified that [it doesn't archive submodules](https://stackoverflow.com/questions/5482544/how-can-i-use-git-archive-to-include-submodules-from-a-bare-repository) and thus will need to have a workaround.

I'll see what I can do.

**Comment by Qix-** [comment]
Also, I highly suggest that support for remote archiving is removed. Git has terrible support for it, and it would require a _lot_ of extra code just to get submodules working. As it stands, there isn't any code that enables any of the features, and `github.com` is explicitly disabled for remote archiving.

I'll make a note of this in the eventual PR and ask the owners of Yarn if it should be removed. I agree with the comment in-source:

```javascript
const supportsArchiveCache: {[key: string]: boolean} = map({
  'github.com': false, // not support, doubt they will ever support it
});
```

**Comment by andreimc** [comment]
is there any fix for this ?

**Comment by pgkehle** [comment]
We are officially waiting on this too as an offshoot from the https://github.com/oracle/node-oracledb v2.0 will need this to work.

**Comment by pgkehle** [comment]
@brunolemos No I do not.  I work at NC State. But we are using oracle servers, some of which still use the legacy LONG RAW type, and it's a need for what my department is developing.    We do not have the bandwidth to look into making that install issue work.  😞

**Comment by jokeyrhyme** [comment]
As a work around, you could remove git dependencies from your package.json, and have a postinstall hook script
The script could either fallback to `npm install` just for the git dependencies, or it could use `git` directly to clone them into node_modules, and then descends into each one and runs `yarn` or `npm install` (if they themselves have dependencies)

**Comment by pgkehle** [comment]
@jokeyrhyme yes, that is what we are doing currently. See https://github.com/oracle/node-oracledb/issues/794#issuecomment-344995673, in case someone needs an example of how to do that.

**Comment by TheAifam5** [comment]
What's the status? This still does not work :(

**Comment by Qix-** [comment]
The status is nobody has submitted a PR and the ticket is still open, @TheAifam5.

**Comment by nitaking** [comment]
>In package.json of a module being fetched, add a following script
`"postinstall": "grep url .gitmodules | sed 's/.*= //' | while read url; do git clone $url; done"`

If deal with an empty submodule folder with `git clone`, a permission error has occurred and you can not` yarn add` it.

what should I do?

**Comment by nitaking** [comment]
I solved it, using a symbolic link.

```
repositoryA
├── repositoryC(submodule)
│     └── hogehoge
│
└── node_modules/
    └── repositoryB/
        └── repositoryC(symbolic link)
             └── (hogehoge)
```

This is only a temporary solution, but ... https://qiita.com/nitaking/items/3340955e80144ec90ad6

**Comment by reyalpsirc** [comment]
In my case, to be able to install it on the correct paths of .gitmodules, I needed to set the postinstall on package.json as this:

	"postinstall": "sed -n -e '/path/,/url/p' .gitmodules | sed 'N;s/\\n/\\$$$/' | while IFS= read -r line; do if [[ $line =~ (.*)\\$\\$\\$(.*) ]] ; then path=\"$(echo ${BASH_REMATCH[1]} | sed 's/.*= //')\"; url=\"$(echo ${BASH_REMATCH[2]} | sed 's/.*= //')\"; git clone $url $path; fi done",

**Comment by Shinobu1337** [comment]
Any progress on this?

**Comment by Qix-** [comment]
@Shinobu1337 If there's nothing here, then no. Progress in the OSS community is pretty transparent.

**Comment by john-shine** [comment]
@reyalpsirc not work in my computer, so add some improves to your codes. It should manually replace xxxxx with your packages which contain submodules: 

`"postinstall": "project_dir=$PWD; modules=( 'xxxxx' ); for module in \"${modules[@]}\"; do echo \"Notice: update module ${module}\"; cd \"./node_modules/${module}\"; if [[ -f .gitmodules ]]; then sed -n -e '/path/,/url/p' .gitmodules | sed 'N;s/\\n/\\$$$/' | while IFS= read -r line; do if [[ $line =~ (.*)\\$\\$\\$(.*) ]] ; then path=\"$(echo ${BASH_REMATCH[1]} | sed 's/.*= //')\"; url=\"$(echo ${BASH_REMATCH[2]} | sed 's/.*= //')\"; if [[ -d \"${path}/.git\" ]] ; then echo \"Notice: git pull submodule $(realpath ${path})\"; cd \"${path}\" && git pull; else if [[ -z \"$(ls -A ${path})\" ]]; then echo \"Notice: git clone submodule $(realpath ${path})\"; rm -rf \"${path}\"; git clone $url $path; else echo \"Notice: $(realpath ${path}) submodule folder is not empty, try to remove it before clone.\"; fi; fi; else echo \"Notice: module ${module} is not contain submodules.\"; fi; done; else echo \"Notice: module ${module} is not contain submodules.\"; fi; cd \"${project_dir}\"; done;"`

Test on macOS bash

**Comment by EAlexis92** [comment]
Not fixed yet? Ok time to move on to something else

**Comment by Superlokkus** [comment]
This is very disappointing, especially since npm fixed this years ago. Now I can't not use any dependencies without a dummy package.json added. napa would be a workaround if I wouldn't have a install step (cmake-js), since you can't not portable chain install steps with npm. All post install steps here fail due to not being portable on windows.
